### PR TITLE
Format Learn how a stack works challenge

### PR DESCRIPTION
**Pull request by erictleung**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.
- [ ] Closes currently open issue (replace XXXX with an issue no): Closes #XXXX

#### Description

- Improve readability of challenge
- Spell out JavaScript abbreviations
- Apply appropriate dfn and code tags
- Change from single quotes to double quotes
- Add periods to end of challenge tests

Related #12757

### CODE REVIEW

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 1]
```diff
@@ -45,23 +45,25 @@
       "id": "587d8250367417b2b2512c5e",
       "title": "Learn how a Stack Works",
       "description": [
-        "You are probably familiar with stack of books on your table. You might have used the undo feature in your computer. You are also used to hitting the back button on your phone to go back to the previous view in your app.",
+        "You are probably familiar with stack of books on your table. You might have used the undo feature in your computer. You are also probably used to hitting the back button on your phone to go back to the previous view in your app.",
```
Comment (dhcodes) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: I recommend changing "You might have used the undo feature in your computer." to "You have likely used the Undo feature of a text editor".

"In your computer" seems generic.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 2]
```diff
@@ -45,23 +45,25 @@
...
-        "Instructions",
-        "Here we have a stack of homework assignments represented as an array: BIO12 is at the base, and PSY44 is at the top of the stack.",
-        "Remove the element on top of the stack, and Add CS50."
+        "The topmost book in the stack was the one that was put there last. If you remove that book from your stack's top, you would expose the book that was put there before the last book and so on.",
+        "If you think about it, in all the above examples, you are getting <dfn>Last-In-First-Out</dfn> type of service. We will try to mimic this with our code.",
+        "Oh, and did I mention such a data storage scheme is called a <dfn>Stack</dfn>! In particular, we would have to implement the <code>push()</code> method that pushes JavaScript objects at the top of the stack; and <code>pop()</code> method, that removes the JavaScript object that's at the top of the stack at the current moment.",
```
Comment (dhcodes) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Recommend changing:
>"Oh, and did I mention such a"

to:
 
>"This". 

and removing the `!`

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 3]
```diff
@@ -45,23 +45,25 @@
...
+        "<hr>",
+        "Here we have a stack of homework assignments represented as an array: <code>\"BIO12\"</code> is at the base, and <code>\"PSY44\"</code> is at the top of the stack.",
+        "Create a stack data storage scheme with JavaScript arrays and functions that we always get back first what we put there last.",
```
Comment (dhcodes) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: I recommend changing:
>"Create a stack data storage scheme with JavaScript arrays and functions..."

to:

>Write a function that accepts an array and a new item and removes the last item of the array and replaces it with the new item. Then call your function using `homeworkStack` as the array and `"CS50"` as the new item that will replace `"PSY44"`.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json (hunk 1 above)
Comment (dhcodes) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: I recommend changing:
>"You might have used the undo feature in your computer." 

to

>"You have likely used the Undo feature of a text editor".

"In your computer" seems generic.

**Review by dhcodes — CHANGES_REQUESTED** [review]
Please see comments.

**Comment by erictleung** [comment]
@dhcodes thanks for the feedback! I've been bogged down with other stuff the past few days. I'll get to this by sometime this weekend.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json (hunk 3 above)
Comment (erictleung) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: This challenge only requires you to use `pop` and `push`. You don't need a function to pass the challenge. The tests only look at `homeworkStack`. I can change it something like

> Modify the given array and treat it like a `stack` using JavaScript methods. Remove the top element `"BIO12"` from the stack. Then add `"CS50"` to be the new top element of the stack. 

But now that I look at it, the unit tests need a bit of work. You can pass the challenge if you just set `homeworkStack` to the correct array elements.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 4]
```diff
@@ -45,23 +45,24 @@
...
+        "If you think about it, in all the above examples, you are getting <dfn>Last-In-First-Out</dfn> type of service. We will try to mimic this with our code.",
+        "This data storage scheme is called a <dfn>Stack</dfn>. In particular, we would have to implement the <code>push()</code> method that pushes JavaScript objects at the top of the stack; and <code>pop()</code> method, that removes the JavaScript object that's at the top of the stack at the current moment.",
+        "<hr>",
+        "Here we have a stack of homework assignments represented as an array: <code>\"BIO12\"</code> is at the base, and <code>\"PSY44\"</code> is at the top of the stack.",
+        "Modify the given array and treat it like a <code>stack</code> using the JavaScript methods mentioned above. Remove the top element <code>\"BIO12\"</code> from the stack. Then add <code>\"CS50\"</code> to be the new top element of the stack."
       ],
       "challengeSeed": [
-        "var homeworkStack = ['BIO12','HIS80','MAT122','PSY44'];",
-        "// Only change code below this line"
+        "var homeworkStack = [\"BIO12\",\"HIS80\",\"MAT122\",\"PSY44\"];",
```
Comment (Greenheart) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: This is rather unserious, but why the double quotes?

By using single quotes, you can save campers from downloading the extra bytes for the escaping backslashes - just saying... :wink:

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 5]
```diff
@@ -45,23 +45,24 @@
...
+        "// Only change code below this line",
+        ""
       ],
       "tests": [
-        "assert(homeworkStack.length === 4, 'message: <code>homeworkStack</code> should only contain 4 elements');",
-        "assert(homeworkStack[3] === 'CS50', 'message: The last element in <code>homeworkStack</code> should be CS50');",
-        "assert(homeworkStack.indexOf('PSY44') === -1, 'message: <code>homeworkStack</code> should not contain PSY44');"
+        "assert(homeworkStack.length === 4, 'message: <code>homeworkStack</code> should only contain 4 elements.');",
+        "assert(homeworkStack[3] === 'CS50', 'message: The last element in <code>homeworkStack</code> should be <code>\"CS50\"</code>.');",
+        "assert(homeworkStack.indexOf('PSY44') === -1, 'message: <code>homeworkStack</code> should not contain <code>\"PSY44\"</code>.');"
```
Comment (Greenheart) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: This still allows the challenge to be passed

```js
var homeworkStack = ["BIO12","HIS80","MAT122","PSY44"];
// Only change code below this line

homeworkStack = ['BIO12', 'HIS80', 'MAT122', 'CS50'];
```

I see two approaches to making the tests more robust:
1. Use a regex to verify that the first line of the challengeSeed is intact, and that the `code` only contains one instance of the `=` character. **This is probably the best method**.
2. Alternatively we can verify that the array methods have actually been used.

What do you think?

**Review by Greenheart — CHANGES_REQUESTED** [review]
@erictleung Nice work! :blush:

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json (hunk 4 above)
Comment (erictleung) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Haha I've just seen only double quotes in seed codes, so I'm just trying to be consistent :smile:

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json (hunk 5 above)
Comment (erictleung) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: @Greenheart I've added this issue to address the insufficiencies in the challenge tests https://github.com/freeCodeCamp/freeCodeCamp/issues/13459. I can go ahead and add better tests here but this PR was just on formatting.

Comment (Greenheart) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: @erictleung Right, then I'll merge this :blush:

**Pull request merged**